src/backend/env/invite.rs

```diff
@@ -0,0 +1,143 @@
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeSet;
+
+#[derive(Default, Serialize, Deserialize, Clone)]
+pub struct Invite {
+    pub credits: Credits,
+    pub credits_per_user: Credits,
+    pub joined_user_ids: BTreeSet<UserId>,
+    pub realm_id: Option<RealmId>,
+    pub inviter_user_id: UserId,
+}
+
+impl Invite {
+    pub fn new(
+        credits: Credits,
+        credits_per_user: Credits,
+        realm_id: Option<RealmId>,
+        inviter_user_id: UserId,
+    ) -> Self {
+        // Convert empty realm_id to None
+        let converted_relm_id: Option<RealmId> = realm_id.filter(|id| !id.is_empty());
+
+        Self {
+            credits,
+            credits_per_user,
+            joined_user_ids: BTreeSet::new(),
+            realm_id: converted_relm_id,
+            inviter_user_id,
+        }
+    }
+
+    pub fn consume(&mut self, joined_user_id: UserId) -> Result<(), String> {
+        if self.joined_user_ids.contains(&joined_user_id) {
+            return Err("User already credited".into());
+        }
+
+        let new_credits = self
+            .credits
+            .checked_sub(self.credits_per_user)
+            .ok_or("Invite credits too low")?;
+        self.credits = new_credits;
+
+        self.joined_user_ids.insert(joined_user_id);
+
+        Ok(())
+    }
+
+    pub fn update(
+        &mut self,
+        credits: Option<Credits>,
+        realm_id: Option<RealmId>,
+        user_id: UserId,
+    ) -> Result<(), String> {
+        if self.inviter_user_id != user_id {
+            return Err("Owner does not match".into());
+        }
+
+        if let Some(new_credits) = credits {
+            if new_credits % self.credits_per_user != 0 {
+                return Err(format!(
+                    "Credits per user are not a multiple of credits {} {}",
+                    new_credits, self.credits_per_user
+                ));
+            }
+
+            // Protect against creating invite and setting to 0 without usage
+            if new_credits == 0 && self.joined_user_ids.is_empty() {
+                return Err("Cannot set credits to 0 as it has never been used".into());
+            }
+
+            self.credits = new_credits;
+        }
+
+        if let Some(id) = realm_id {
+            if id.is_empty() {
+                self.realm_id = None;
+            } else {
+                self.realm_id = Some(id);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub fn invites_by_principal(
+    state: &State,
+    principal: Principal,
+) -> Box<impl Iterator<Item = (&String, &Invite)>> {
+    let invites = state
+        .principal_to_user(principal)
+        .ok_or("Principal not found")
+        .map(|user| user.id)
+        .map(|user_id| {
+            state
+                .invite_codes
+                .iter()
+                .filter(move |(_, invite)| invite.inviter_user_id == user_id)
+        })
+        .expect("Failed to filter invites");
+
+    Box::new(invites)
+}
```

Comment:
How about avoiding panicking?

```suggestion
pub fn invites_by_principal(
    state: &State,
    principal: Principal,
) -> Box<dyn Iterator<Item = (&'_ String, &'_ Invite)> + '_> {
    match state.principal_to_user(principal).map(|user| {
        state
            .invite_codes
            .iter()
            .filter(move |(_, invite)| invite.inviter_user_id == user.id)
    }) {
        Some(iter) => Box::new(iter),
        _ => Box::new(std::iter::empty()),
    }
}
```

Reply:
Much better, thanks!